### DISCUSSION TITLE
Angular Universal support

### DISCUSSION THREAD

**Discussion by pavankjadda** [stevermeister/ngx-cookie-service#166] [General]
### Is your feature request related to a problem? Please describe

No

### Describe the solution you'd like

Support Angular Universal (SSR) 

### Describe alternatives you've considered

No

**Comment by pavankjadda** [comment]
@stevermeister Just to summarize from #1 

We don't have to store cookies in Node or any other server. We store them in browser as usual but we have to make sure `document` is available before accessing it. Since we are already injecting `document` based on [Angular docs](https://angular.io/guide/universal#working-around-the-browser-apis), the library works as expected

 I created [sample repo](https://github.com/pavankjadda/angular-ssr-docker) that demonstrates this. It works as expected. Clone the project and run as per instructions in Readme.md file

**Comment by pavankjadda** [comment]
@stevermeister I found another issue when using this library in SSR. Authentication guards fail when page loads initially because document is not available. 
1. When home page loads, home page guard fails as it looks for cookies in ngx cookie service which depends `document`
2. Then Guard redirects to Login page
3. When `document` is available, the guard condition passes and redirects back to home page

Even though user goes to original requested page, but he sees redirection to login page, which is not ideal. 

I believe this caused by the [code where are looking](https://github.com/stevermeister/ngx-cookie-service/blob/master/projects/ngx-cookie-service/src/lib/cookie.service.ts#L19) checking `isPlatformBrowser` not `isPlatformServer`. I will do some research on this.

**Comment by stevermeister** [comment]
@pavankjadda no I'm on a business trip and don't have time to dive deep into it. But it's a really interesting topic. Maybe we can even set up a call for next week?

**Comment by pavankjadda** [comment]
@stevermeister How do you wanna meet? We have [Discord server](https://discord.com/channels/873021904708059177/873021905463029852) if you like to meet there. Or I can send WebEx link, let me know.

**Comment by stevermeister** [comment]
@pavankjadda what would be a convenient day/time for you? 
I'm new to Discord, not sure what I'm doing wrong, but your link brings me to this message
<img width="529" alt="Screenshot 2021-10-18 at 18 19 10" src="https://user-images.githubusercontent.com/1526680/137770010-eb14f853-eb63-481f-abfd-a4992689bdfa.png">

**Reply by pavankjadda** [reply]
> @pavankjadda what would be a convenient day/time for you? 
> 
> I'm new to Discord, not sure what I'm doing wrong, but your link brings me to this message
> 
> <img width="529" alt="Screenshot 2021-10-18 at 18 19 10" src="https://user-images.githubusercontent.com/1526680/137770010-eb14f853-eb63-481f-abfd-a4992689bdfa.png">

Try this link https://discord.gg/ta8swVRWNF

**Comment by stevermeister** [comment]
@pavankjadda I had a bit of time to think about it. 

I see 2 questions that we need to solve here: 
- how to inject cookies 
- structure

# how to inject cookies 

I liked the idea that you suggested: inject cookies via ngExpressEngine.
created simple app here to check - https://github.com/stevermeister/ssr-coookie
From our library, we need to provide **InjectionsToken** that should be used in express.


# structure

The first thought was to create cookie-client, cookie-server separate classes and after factory to create instance by condition. 
But then I realized that it's probably an overkill and we only need to patch 2 methods. 
What do you think?

**Reply by pavankjadda** [reply]
I agree with approach. But when it comes to implementation, we need to make sure the service is injected in `app.server.module.ts` and then modify the request to include cookies. See my reply to your question on Discord

**Comment by stevermeister** [comment]
created initial PR for it - https://github.com/stevermeister/ngx-cookie-service/pull/171/files

**Comment by pavankjadda** [comment]
@stevermeister I have a workaround to use current version of the library without any changes. See the [StackOverFlow issue](https://stackoverflow.com/a/69742115/9244861) for details.

**Reply by stevermeister** [reply]
@pavankjadda basically, you return TRUE if it's a server-side? :)

**Reply by pavankjadda** [reply]
> @pavankjadda basically, you return TRUE if it's a server-side? :)

Yeah. But it does not work as expected.

**Reply by stevermeister** [reply]
well, for this case you just skip the auth for the server-side, that is not correct

**Comment by martijnmelchers** [comment]
@pavankjadda I see the SSR merge has been reverted again, why was this and any idea when it'll be returning?

**Reply by pavankjadda** [reply]
Yes. It broke existing applications that doesn't use SSR. It will come back soon.
